gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactionTask.java

```diff
@@ -0,0 +1,341 @@
+    job.setMapOutputKeyClass(AvroKey.class);
+    job.setMapOutputValueClass(AvroValue.class);
+  }
+
+  /**
+   * mimic {@link MRCompactorAvroKeyDedupJobRunner#configureReducer(Job)}
+   */
+  protected void configureReducer(Job job) throws IOException {
+    job.setOutputFormatClass(AvroKeyCompactorOutputFormat.class);
+    job.setReducerClass(AvroKeyDedupReducer.class);
+    job.setOutputKeyClass(AvroKey.class);
+    job.setOutputValueClass(NullWritable.class);
+    setNumberOfReducers(job);
+  }
+
+  /**
+   * mimic {@link MRCompactorAvroKeyDedupJobRunner#setNumberOfReducers(Job)}
+   */
+  protected void setNumberOfReducers(Job job) throws IOException {
+
+    // get input size
+    long inputSize = 0;
+    for (Path inputPath : this.mrInputPathes) {
+      inputSize += this.fs.getContentSummary(inputPath).getLength();
+    }
+
+    // get target file size
+    long targetFileSize = this.state.getPropAsLong(MRCompactorAvroKeyDedupJobRunner.COMPACTION_JOB_TARGET_OUTPUT_FILE_SIZE,
+            MRCompactorAvroKeyDedupJobRunner.DEFAULT_COMPACTION_JOB_TARGET_OUTPUT_FILE_SIZE);
+
+    // get max reducers
+    int maxNumReducers = state.getPropAsInt(MRCompactorAvroKeyDedupJobRunner.COMPACTION_JOB_MAX_NUM_REDUCERS,
+            MRCompactorAvroKeyDedupJobRunner.DEFAULT_COMPACTION_JOB_MAX_NUM_REDUCERS);
+
+    int numReducers = Math.min(Ints.checkedCast(inputSize / targetFileSize) + 1, maxNumReducers);
+
+    // get use prime reducers
+    boolean usePrimeReducers = state.getPropAsBoolean(MRCompactorAvroKeyDedupJobRunner.COMPACTION_JOB_USE_PRIME_REDUCERS,
+            MRCompactorAvroKeyDedupJobRunner.DEFAULT_COMPACTION_JOB_USE_PRIME_REDUCERS);
+
+    if (usePrimeReducers && numReducers != 1) {
+      numReducers = Primes.nextPrime(numReducers);
+    }
+    job.setNumReduceTasks(numReducers);
+  }
+
+  /**
+   * Concatenate multiple directories into one path
+   */
+  private Path concatPaths (String a, String b, String c, String d) {
+    return new Path (a, new Path (b, new Path (c, new Path (d))));
+  }
+
+  /**
+   * mimic {@link MRCompactorAvroKeyDedupJobRunner#configureInputAndOutputPaths(Job)}
+   */
+  protected void configureInputAndOutputPaths(Job job) throws IOException {
+    for (Path path: mrInputPathes) {
+      FileInputFormat.addInputPath(job, path);
+    }
+
+    String mrOutputBase = this.state.getProp(MRCompactor.COMPACTION_TMP_DEST_DIR);
+
+    CompactionParser.CompactionParserResult rst = this.suite.getParser().parse(this.partition);
+    this.mrOutput = concatPaths (mrOutputBase, rst.getDatasetName(), rst.getDstSubDir(), rst.getTimeString());
+
+    log.debug ("Cleanup temp MR output directory: " + mrOutput);
+    this.fs.delete(mrOutput, true);
+    FileOutputFormat.setOutputPath(job, mrOutput);
+  }
+
+  @Override
+  public void run() {
+    for (CompactionVerifier verifier: getRequiredVerifier()) {
+      if (!verifier.verify(this.partition)) {
+        return;
+      }
+    }
+    super.run();
+    onMRJobCompletion();
```

Comment:
I replaced this method with a list of CompactionCompeteAction instances. I think now it's better to keep them in a compaction aware context.